### PR TITLE
Add Trino 431 release notes

### LINKED ISSUE #19475: Release notes for 431

### PR DESCRIPTION
**Pull request by mosabua**
## Description

Assemble the release notes for the 431 release notes.

## Additional context and related issues

Fixes #19475 

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

## Verification for each pull request

Format: maintainer, PR/issue number, ✅ / ❌ rn ✅ / ❌ docs
✅ rn - release note added and verified, or assessed to be not necessary, set to ❌ rn before completion
✅ docs - need for docs assessed and merged, or assessed to be not necessary, set to ❌ docs before completion

## 20 Oct 2023

* #19385 ✅ rn ✅ docs
* #19474 ✅ rn ✅ docs

## 21 Oct 2023

* #19470 ✅ rn ✅ docs
* #19476 ✅ rn ✅ docs
* #19480 ✅ rn ✅ docs

## 22 Oct 2023

* #19390 ✅ rn ✅ docs

## 23 Oct 2023

* #19439 ✅ rn ✅ docs
* #19413 ✅ rn ✅ docs
* #19486 ✅ rn ✅ docs
* #19022 ✅ rn ✅ docs, see #19502 
* #19340 ✅ rn ✅ docs
* #19445 ✅ rn ✅ docs

## 24 Oct 2023

* #19505 ✅ rn ✅ docs
* #19515 ✅ rn ✅ docs
* #19508 ✅ rn ✅ docs
* #19494 ✅ rn ✅ docs
* #19518 ✅ rn ✅ docs
* #19434 ✅ rn ✅ docs
* #19015 ✅ rn ✅ docs
* #19502 ✅ rn ✅ docs
* #19521 ✅ rn ❌ ? docs
* #19519 ??? not sure if we should add rn ✅ docs

## 25 Oct 2023

* #13681 ✅ rn ✅ docs
* #19525 ✅ rn ✅ docs
* #19534 ✅ rn ✅ docs
* #19496 ✅ rn ✅ docs
* #19308 ✅ rn ✅ docs
* #18854 ✅ rn ✅ docs
* #19487 ✅ rn ✅ docs
* #17635 ✅ rn ✅ docs
* #19488 ✅ rn ✅ docs
* #19463 ✅ rn ✅ docs

## 26 Oct 2023

* #18464 ✅ rn ✅ docs
* #19528 ✅ rn ✅ docs
* #19509 ✅ rn ✅ docs
* #19478 ✅ rn ✅ docs
* #19530 ✅ rn ✅ docs
* #19541 ✅ rn ✅ docs

## 27 Oct 2023

* #19275 ✅ rn ✅ docs
* #19537 ✅ rn ✅ docs
* #19542 ✅ rn ✅ docs

### CODE REVIEW

**Comment by mosabua** [comment]
@dain  please confirm the reworded entries in the release notes for #19385 and #19480

**Comment by mosabua** [comment]
@wendigo please confirm that there are really no user facing updates for docs and release notes necessary for #19470

**Comment by mosabua** [comment]
@electrum please confirm the reworded entries for #19390

File: docs/src/main/sphinx/release/release-431.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+## SingleStore connector
+
+## SQL Server connector
+
+## TPC-H connector
+
+## SPI
+
+* Add `ValueBlock` abstraction along with `VALUE_BLOCK_POSITION` and
+  `VALUE_BLOCK_POSITION_NOT_NULL` calling convention. ({issue}`19385`)
```
Comment (dain) [inline docs/src/main/sphinx/release/release-431.md:107]:
```suggestion
  `VALUE_BLOCK_POSITION_NOT_NULL` calling conventions. ({issue}`19385`)
```

File: docs/src/main/sphinx/release/release-431.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+* Require a separate block position for each argument for aggregation functions
+  for each argument. ({issue}`19385`)
```
Comment (dain) [inline docs/src/main/sphinx/release/release-431.md:109]:
```suggestion
* Require a separate block position for each argument of aggregation functions. ({issue}`19385`)
```

**Comment by mosabua** [comment]
> @mosabua nope, these are not customer-facing changes.

Thanks for confirming @wendigo

File: docs/src/main/sphinx/release/release-431.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Release 431 (25 Oct 2023)
+
+## General
+
+* Add support for `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-431.md:5]:
```suggestion
* Add support for the `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)
```

File: docs/src/main/sphinx/release/release-431.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+* Disallow a null offset in {func}`lead` and {func}`lag`. ({issue}`19003`)
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-431.md:6]:
```suggestion
* Disallow an offset of `null` in the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
```
I think this is a little clearer.

File: docs/src/main/sphinx/release/release-431.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+## JDBC driver
+
+* Add a new JDBC property `legacyPreparedStatements` that allows to control
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-431.md:10]:
```suggestion
* Add a new JDBC property `legacyPreparedStatements` that controls
```

File: docs/src/main/sphinx/release/release-431.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  whether `EXECUTE IMMEDIATE` is used to execute prepared statements. ({issue}`18957`)
+
+## Delta Lake connector
+
+* Replace the Hive metastore configuration property `hive.metastore-timeout`
```
Comment (colebow) [inline docs/src/main/sphinx/release/release-431.md:15]:
```suggestion
* Replace the `hive.metastore-timeout` Hive metastore configuration property
```
Re-order for consistency w/other notes and flow.
etc. for the other notes.

**Review by colebow — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/release/release-431.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+
+## Hive connector
+
+* Replace the Hive metastore configuration property `hive.metastore-timeout`
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+* Mask Hive S3 proxy password value in logs. ({issue}`19519`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-431.md:28]: Do these two have to go into Iceberg, Delta Lake and Hudi as well @wendigo ?

**Review by colebow — APPROVED** [review]

File: docs/src/main/sphinx/release/release-431.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
+# Release 431 (25 Oct 2023)
+
+## General
+
+* Add support for the `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-431.md:5]: `CREATE TABLE` already exists. Technically, this adds `REPLACE TABLE`. It can only be used in combination with `CREATE TABLE`, in the form of `CREATE OR REPLACE TABLE`.

File: docs/src/main/sphinx/release/release-431.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+* Disallow a offset of `null` in the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
```
Comment (dain) [inline docs/src/main/sphinx/release/release-431.md:6]: `an`, but this sounds off to me.  Maybe:
```suggestion
* Disallow `null` value for the offset of the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
```

File: docs/src/main/sphinx/release/release-431.md (hunk 7 above)
Comment (dain) [reply docs/src/main/sphinx/release/release-431.md:28]: I'm pretty sure they are used there also... the file system code is shared across all of them

File: docs/src/main/sphinx/release/release-431.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+
+## Hive connector
+
+* Replace the `hive.metastore-timeout` Hive metastore configuration property
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+* Mask Hive S3 proxy password value in logs. ({issue}`19519`)
+* Mask key store and trust store password values used for Hive metastore access
+  in logs. ({issue}`19519`)
```
Comment (dain) [inline docs/src/main/sphinx/release/release-431.md:30]: same

File: docs/src/main/sphinx/release/release-431.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+* Add support for the `CREATE OR REPLACE TABLE` statement. ({issue}`13180`)
+* Disallow a offset of `null` in the {func}`lead` and {func}`lag` functions. ({issue}`19003`)
+
+## Web UI
+
+* Mask shared secret value in logs. ({issue}`19519`)
+
+## JDBC driver
+
+* Add a new JDBC property `legacyPreparedStatements` that controls whether
```
Comment (martint) [inline docs/src/main/sphinx/release/release-431.md:14]: This is too low level. What's the user visible impact? (something like "improve query latency by reducing number of roundtrips when executing prepared statements")

File: docs/src/main/sphinx/release/release-431.md (hunk 9 above)
Comment (martint) [reply docs/src/main/sphinx/release/release-431.md:6]: Disallow a `null` offset ...

File: docs/src/main/sphinx/release/release-431.md (hunk 8 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:5]: I know .. how do we say that ... 

Add support for `OR REPLACE` with `CREATE TABLE` statements.  

??

File: docs/src/main/sphinx/release/release-431.md (hunk 11 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:14]: Yeah .. well.. it does do that since its off by default ..

File: docs/src/main/sphinx/release/release-431.md (hunk 9 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:6]: Adjusted

File: docs/src/main/sphinx/release/release-431.md [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+
+## Hive connector
+
+* Add support for [SQL routine management](sql-routine-management). ({issue}`19308`)
+* Replace the `hive.metastore-timeout` Hive metastore configuration property
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+* Improve support for concurrent table statistics updates in Glue. ({issue}`19463`)
```
Comment (mosabua) [inline docs/src/main/sphinx/release/release-431.md:33]: fyi @findepi and @findinpath .. I reworded to this. Also does this apply to Iceberg, Delta Lake, or Hudi connectors as well?

Comment (findinpath) [reply docs/src/main/sphinx/release/release-431.md:33]: To my knowledge it applies only to Hive.

Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:33]: Thanks

File: docs/src/main/sphinx/release/release-431.md [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+
+## Hive connector
+
+* Add support for [SQL routine management](sql-routine-management). ({issue}`19308`)
+* Replace the `hive.metastore-timeout` Hive metastore configuration property
+  with the `hive.metastore.thrift.client.connect-timeout` and
+  `hive.metastore.thrift.client.read-timeout` properties. ({issue}`19390`)
+* Cast char fields, if necessary, to varchar type in Hive view translations. ({issue}`18337`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-431.md:34]: That's the implementation detail. What problem is this solving?

File: docs/src/main/sphinx/release/release-431.md [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+* Improve support for concurrent table statistics updates in Glue. ({issue}`19463`)
```
Comment (martint) [inline docs/src/main/sphinx/release/release-431.md:35]: "Improve support for concurrent updates of table statistics in Glue."

File: docs/src/main/sphinx/release/release-431.md [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+* Improve performance of queries with short running splits. ({issue}`19487`)
+
+## Security
+
+* Support defining rules for procedures in file-based access control. ({issue}`19416`)
+* Mask additional sensitive values in log files. ({issue}`19519`)
+
+## JDBC driver
+
+* Allow configuration of executing statements with the legacy
```
Comment (martint) [inline docs/src/main/sphinx/release/release-431.md:18]: This doesn't say what's the impact. I would rephrase as something along the lines of:

> Improve latency for prepared statements for Trino versions that support `EXECUTE IMMEDIATE` when the `explicitPrepare` parameter to is set to `false`.

Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:18]: Updated

File: docs/src/main/sphinx/release/release-431.md (hunk 13 above)
Comment (mosabua) [reply docs/src/main/sphinx/release/release-431.md:34]: Prevents query translations failures.. updated

**Review by martint — APPROVED** [review]
